toolchain/parse/handle_expr.cpp

```diff
@@ -208,11 +208,16 @@ auto HandleExprInPostfix(Context& context) -> void {
       break;
     }
     default: {
+      CARBON_DIAGNOSTIC(ExpectedExpr, Error, "expected expression");
+      context.emitter().Emit(*context.position(), ExpectedExpr);
+
+      // Fallthrough to the error token case -- we don't need to diagnose those.
+      [[fallthrough]];
+    }
+    case Lex::TokenKind::Error: {
```

Comment:
I'm not used to seeing a non-last `default` case (enough that I wouldn't have expected this to work), but maybe others are more versed with that structure? Had you considered writing this as an `if` instead of `fallthrough`, e.g.: `default: if (token_kind != Lex::TokenKind::Error) { ...Emit... }`

Reply:
:shrug: The structure didn't give me any pause, but we've established that I'm not necessarily representative there...

I had tried both ways of writing it, but the switch seemed better. Using an if is a bit awkward as you have to dig the kind back out and then re-test it. As we're already testing it for the switch, and there is a natural fallthrough structure, it seemed clean to use that instead.

Reply:
> 🤷 The structure didn't give me any pause, but we've established that I'm not necessarily representative there...
> 
> I had tried both ways of writing it, but the switch seemed better. Using an if is a bit awkward as you have to dig the kind back out and then re-test it. As we're already testing it for the switch, and there is a natural fallthrough structure, it seemed clean to use that instead.

What do you mean by digging it back out? Couldn't you add `auto token_kind =` in the `switch` statement?

Reply:
I mean, yes, I could also store it.

But it still ends up with an awkward thing where every other value is handled by a case, but this one isn't.

If you feel strongly that using fallthrough isn't OK here, I can change it I guess? Didn't seem like a big thing either way.

Reply:
I'm really having trouble. [#style](https://discord.com/channels/655572317891461132/821113559755784242/1299399143365804044) to see if it's just me.

Reply:
No stress, this really wasn't an important one, I'll switch it to the other form. Was really just trying to understand if it was just surprise or causing more trouble. It's slightly awkward to use the if, but as you say, very slight so seems easily outweighed given this isn't working at all.

Reply:
Back to an `if`, with the suggested formulation.